### PR TITLE
Housekeeping cgroup: Add full-pcpu-only support

### PR DESCRIPTION
**Pull request by RamLavi**
This PR updates the Housekeeping cgroup to support multiple CPUs.
When setting the VM to full-pcpu-only, it is possible to get a full core - containing multiple CPUs - to the emulator thread [0].
This PR assigns these CPUs to the Housekeeping cgroup, so that they can be used on the emulatorThread processes. 

Note: this PR implements the second phase of the https://github.com/kubevirt/community/pull/247.

[0] https://github.com/kubevirt/kubevirt/pull/10593

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
Support multiple CPUs in Housekeeping cgroup
```

### CODE REVIEW

**Comment by RamLavi** [comment]
Manually checked CPU hotplug , moving from: {1 Cores, 1 sockets} to {1 Cores, 2 sockets}.
New virt-launcher pod changed number of CPUs to 4 as expected.

**Comment by RamLavi** [comment]
> Manually checked CPU hotplug , moving from: {1 Cores, 1 sockets} to {1 Cores, 2 sockets}. New virt-launcher pod changed number of CPUs to 4 as expected.

@vladikr - Do you think we need to add an e2e scenario for CPU-hotplug? 
Note that if we want to check {**2** Cores, 1 sockets} to {2 Cores, 2 sockets} - then the current CPU allocation on  KUBEVIRTCI nodes (=6) won't be enough..

**Comment by RamLavi** [comment]
> > Manually checked CPU hotplug , moving from: {1 Cores, 1 sockets} to {1 Cores, 2 sockets}. New virt-launcher pod changed number of CPUs to 4 as expected.
> 
> @vladikr - Do you think we need to add an e2e scenario for CPU-hotplug? Note that if we want to check {**2** Cores, 1 sockets} to {2 Cores, 2 sockets} - then the current CPU allocation on KUBEVIRTCI nodes (=6) won't be enough..

Issued a PR to increase the CPUs to 8 - let's see if we get any objections on it. https://github.com/kubevirt/kubevirtci/pull/1099

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go [hunk 1]
```diff
@@ -454,18 +454,38 @@ func AdjustDomainForTopologyAndCPUSet(domain *api.Domain, vmi *v12.VirtualMachin
...
+			if err != nil {
+				e := fmt.Errorf("no CPU allocated for the emulation thread: %v", err)
+				log.Log.Reason(e).Error("failed to format emulation thread pin")
+				return e
+			}
+			emulatorThreads = append(emulatorThreads, availableThread)
 		}
-		appendDomainEmulatorThreadPin(domain, emulatorThread)
+
+		emulatorThreadsCPUSet = convertCPUListToCPUSet(emulatorThreads)
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:484]: perhaps this could just stay a slice? it's expected as a slice everywhere including on pkg/virt-handler/vm.go:2835

**Review by vladikr — COMMENTED** [review]
overall, looks good.
just a few minor comments inline.

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go (hunk 1 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:484]: as discussed offline, on virt-hanlder it is taken from the domain - where it is set as a CPUSet then converted to a slice. 
This change on virt-launcher only aims to remove one redundant convert to CPUSet.

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go [hunk 2]
```diff
@@ -454,18 +454,38 @@ func AdjustDomainForTopologyAndCPUSet(domain *api.Domain, vmi *v12.VirtualMachin
...
+	var emulatorThreads []uint32
+	var emulatorThreadsCPUSet string
 	if vmi.Spec.Domain.CPU.IsolateEmulatorThread {
-		emulatorThread, err = cpuPool.FitThread()
-		if err != nil {
-			e := fmt.Errorf("no CPU allocated for the emulation thread: %v", err)
-			log.Log.Reason(e).Error("failed to format emulation thread pin")
-			return e
+		CPUManagerPolicyBetaOption := v12.CPUManagerPolicyBetaOptions(vmi.Annotations[v12.CPUManagerPolicyBetaOptionsAnnotation])
+		if cpuManagerPolicyBetaOptionFullpCPUsOnlyEnabled(CPUManagerPolicyBetaOption) {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:461]: DONE

**Comment by RamLavi** [comment]
As discussed with @vladikr offline - removed the e2e test.

**Review by enp0s3 — COMMENTED** [review]
@RamLavi Thank you! overall looks good! I was thinking, since your scenario assumes full core for housekeeping, why not pin the emulator thread and the IO thread each on a separate CPUset? Another thing is that if you want to share the full core with both the IOthread and the emulator thread, then you can create 2 IOThreads, pin each one of them to a separate cpuset and load-balance the IOthreads among the virtual disks that ask for dedicated IO thread

**Comment by RamLavi** [comment]
> @RamLavi Thank you! overall looks good! I was thinking, since your scenario assumes full core for housekeeping, why not pin the emulator thread and the IO thread each on a separate CPUset? 
Another thing is that if you want to share the full core with both the IOthread and the emulator thread, then you can create 2 IOThreads, pin each one of them to a separate cpuset and load-balance the IOthreads among the virtual disks

Hey @enp0s3, the option to pin the IO Threads to the (now possibly multiple) emulator thread is something that I've discussed with @vladikr (as I'm no expert in IO Threads), and he said that it's not required at this point.

**Comment by RamLavi** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/4257d384cd0438f1c2ea091d27ec57eeae3b37b5..7126f350a68d0104f983630790253df8320f8dad): Introducing unit tests to the emulator thread logic (since we removed the e2e test case).

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go [hunk 3]
```diff
@@ -453,19 +470,12 @@ func AdjustDomainForTopologyAndCPUSet(domain *api.Domain, vmi *v12.VirtualMachin
...
 
-	var emulatorThreads []uint32
 	var emulatorThreadsCPUSet string
 	if vmi.Spec.Domain.CPU.IsolateEmulatorThread {
-		availableThread, err := cpuPool.FitThread()
-		if err != nil {
-			e := fmt.Errorf("no CPU allocated for the emulation thread: %v", err)
-			log.Log.Reason(e).Error("failed to format emulation thread pin")
-			return e
+		if emulatorThreadsCPUSet, err = FormatDomainEmulatorThreadPin(domain, cpuPool); err != nil {
```
Comment (RamLavi) [inline pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go]: suggestion: Move appendDomainEmulatorThreadPin outside of this function to keep the logic separate from the execution

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go [hunk 4]
```diff
@@ -409,17 +410,30 @@ func FormatDomainIOThreadPin(vmi *v12.VirtualMachineInstance, domain *api.Domain
 	return nil
 }
 
-func FormatDomainEmulatorThreadPin(domain *api.Domain, cpuPool VCPUPool) (string, error) {
+func FormatDomainEmulatorThreadPin(domain *api.Domain, vmi *v12.VirtualMachineInstance, cpuPool VCPUPool) (string, error) {
```
Comment (RamLavi) [inline pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:413]: Pass only the annotation

Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:413]: DONE

File: pkg/virt-handler/vm.go [hunk 5]
```diff
@@ -2829,14 +2829,14 @@ func (d *VirtualMachineController) configureHousekeepingCgroup(vmi *v1.VirtualMa
 		return nil
 	}
 
-	hkcpu, err := strconv.Atoi(domain.Spec.CPUTune.EmulatorPin.CPUSet)
+	hkcpus, err := hardware.ParseCPUSetLine(domain.Spec.CPUTune.EmulatorPin.CPUSet, 100)
```
Comment (orelmisan) [inline pkg/virt-handler/vm.go:2832]: nit: As we've discussed offline, please consider naming this magic number that has different values in various calls to `hardware.ParseCPUSetLine`.

File: pkg/virt-launcher/virtwrap/converter/converter_test.go [hunk 6]
```diff
@@ -2029,6 +2030,74 @@ var _ = Describe("Converter", func() {
 		})
 	})
 
+	Context("Correctly handle IsolateEmulatorThread with dedicated cpus", func() {
+		DescribeTable("should fail assigning CPUs to emulatorThread", func(
+			cores uint32, converterContext *ConverterContext, expectedErrorString string) {
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go]: nit:
```suggestion
		DescribeTable("should fail assigning CPUs to emulatorThread", 
			func(cores uint32, converterContext *ConverterContext, expectedErrorString string) {
```

File: pkg/virt-launcher/virtwrap/converter/converter_test.go [hunk 7]
```diff
@@ -2029,6 +2030,74 @@ var _ = Describe("Converter", func() {
...
+							Cpus: []*cmdv1.CPU{
+								{Id: 5}, {Id: 6},
+							},
+						}},
+					},
+				},
+				"no CPU allocated for the emulation thread"),
+		)
+
+		DescribeTable("should succeed assigning CPUs to emulatorThread", func(
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go:2080]: Please consider moving the success scenario before the failure one.

File: pkg/virt-launcher/virtwrap/converter/converter_test.go [hunk 8]
```diff
@@ -2029,6 +2030,74 @@ var _ = Describe("Converter", func() {
...
+			domain.Spec.CPUTune, err = cpuPool.FitCores()
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = vcpu.FormatEmulatorThreadPin(cpuPool)
+			Expect(err).To(MatchError(ContainSubstring(expectedErrorString)))
+		},
+			Entry("when there are not enough CPUs to allocate emulator threads",
+				uint32(2), &ConverterContext{CPUSet: []int{5, 6},
+					AllowEmulation: true,
+					SMBios:         &cmdv1.SMBios{},
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go:2053]: Are there values necessary for the test?

File: pkg/virt-launcher/virtwrap/converter/converter_test.go [hunk 9]
```diff
@@ -2029,6 +2030,74 @@ var _ = Describe("Converter", func() {
...
+			emulatorThreadsCPUSet, err := vcpu.FormatEmulatorThreadPin(cpuPool)
+			Expect(err).ToNot(HaveOccurred())
+			housekeepingCPUs, err := hardware.ParseCPUSetLine(emulatorThreadsCPUSet, 100)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(housekeepingCPUs).To(HaveLen(expectedEmulatorThreads))
+		},
+			Entry("when there is one extra CPU assigned for emulatorThread",
+				uint32(2), &ConverterContext{CPUSet: []int{5, 6, 7},
+					AllowEmulation: true,
+					SMBios:         &cmdv1.SMBios{},
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go:2102]: Are there values necessary for the test?

**Review by orelmisan — COMMENTED** [review]
Over all looks good.
Just a few small questions.

File: pkg/virt-handler/vm.go (hunk 5 above)
Comment (RamLavi) [reply pkg/virt-handler/vm.go:2832]: I suggest doing it in a separate PR for all references of `ParseCPUSetLine`

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 9 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go:2102]: looks like they are not. removing.
DONE

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 7 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go:2080]: DONE

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 8 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go:2053]: looks like they are not. removing.
DONE

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 6 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go]: DONE

**Comment by vladikr** [comment]
Thanks, overall it looks good. 
/approve

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 6 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go]: missed it, Fixed now.

**Review by orelmisan — APPROVED** [review]
Thank you @RamLavi

**Comment by RamLavi** [comment]
/hold

@enp0s3 feel free to remove the hold if you are OK with this change

File: pkg/virt-launcher/virtwrap/converter/converter_test.go [hunk 10]
```diff
@@ -2028,6 +2029,99 @@ var _ = Describe("Converter", func() {
 				"expected number of queues to equal number of requested vCPUs")
 		})
 	})
+
+	Context("Correctly handle IsolateEmulatorThread with dedicated cpus", func() {
+		DescribeTable("should succeed assigning CPUs to emulatorThread",
+			func(cores uint32, converterContext *ConverterContext, CPUManagerPolicyBetaOption v1.CPUManagerPolicyBetaOptions,
```
Comment (enp0s3) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go:2035]: @RamLavi I think it worth to check that we don't overlap with assignment here, that the CPUsets for the emulator threads are not used by VCPU pinning for the guest.

**Comment by enp0s3** [comment]
@RamLavi @vladikr I think the tests are passing only because the test case has the following ratio: the compute container has twice the CPUs than the Guest (guest topology ends up with 2 VCPUs, compute container has 4 CPUsets), in that case the `cpuPool.FitCores()` will indeed manage to return two different VCPU pinning maps. In other case i.e. Guest topology 6 VCPUs and compute container has 8 CPUsets it will fail

**Comment by RamLavi** [comment]
> @RamLavi @vladikr I think the tests are passing only because the test case has the following ratio: the compute container has twice the CPUs than the Guest (guest topology ends up with 2 VCPUs, compute container has 4 CPUsets), in that case the `cpuPool.FitCores()` will indeed manage to return two different VCPU pinning maps. In other case i.e. Guest topology 6 VCPUs and compute container has 8 CPUsets it will fail

you are correct. I need to further understand why this happens.

**Comment by vladikr** [comment]
> > @RamLavi @vladikr I think the tests are passing only because the test case has the following ratio: the compute container has twice the CPUs than the Guest (guest topology ends up with 2 VCPUs, compute container has 4 CPUsets), in that case the `cpuPool.FitCores()` will indeed manage to return two different VCPU pinning maps. In other case i.e. Guest topology 6 VCPUs and compute container has 8 CPUsets it will fail
> 
> you are correct. I need to further understand why this happens.

@RamLavi I think @enp0s3 is right. I missed this.

**Comment by vladikr** [comment]
I think we need a new public function that would allocate a single core. It would simply call `fitCores(coreCount int)` with core count 1

**Comment by RamLavi** [comment]
> I think we need a new public function that would allocate a single core. It would simply call `fitCores(coreCount int)` with core count 1

@vladikr - I think we may have approached this wrong. 
In the first phase [PR](https://github.com/kubevirt/kubevirt/pull/10593) - we are simply asking for another CPU (see [here](https://github.com/kubevirt/kubevirt/blob/5aea2609728247f0fc4f413f50be8ce8ff511118/pkg/virt-controller/services/renderresources.go#L222)) - not demanding a full core. 
The fact that we do get a full core is due to the following reasons (some are out of kubevirt's scope)
- CPU-Manager's full-pcpu-only policy is set.
- if [PAO]([Performance Addon Operator](https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html)) is deployed, then the `single-numa-topology` is set.
- Kubevirt's own CPU pool assignment logic, that always tries to set full cores as much as possible. 

In other words - we shouldn't demand for a full core, but instead ask for another thread.
As we specified in the design proposal - it is the user's responsibility that the cluster is configured in such a way that they will get a full core.

In other words, I think we should use another `fitThread()`. WDYT? 
I will also post the change shortly so you can observe

**Comment by RamLavi** [comment]
manually checking via exec to the virt-launcher pod:
```
sh-5.1$ taskset -p $(pgrep qemu-kvm)
pid 79's current affinity mask: 60
sh-5.1$ cat /proc/$(pgrep qemu-kvm)/status | grep Cpus_allowed_list
Cpus_allowed_list:	5-6
sh-5.1$ virsh dumpxml 1 | grep cpu
Authorization not available. Check if polkit service is running or see debug message for more information.
  <vcpu placement='static' current='4'>16</vcpu>
  <vcpus>
    <vcpu id='0' enabled='yes' hotpluggable='no' order='1'/>
    <vcpu id='1' enabled='yes' hotpluggable='no' order='2'/>
    <vcpu id='2' enabled='yes' hotpluggable='no' order='3'/>
    <vcpu id='3' enabled='yes' hotpluggable='no' order='4'/>
    <vcpu id='4' enabled='no' hotpluggable='yes'/>
    <vcpu id='5' enabled='no' hotpluggable='yes'/>
    <vcpu id='6' enabled='no' hotpluggable='yes'/>
    <vcpu id='7' enabled='no' hotpluggable='yes'/>
    <vcpu id='8' enabled='no' hotpluggable='yes'/>
    <vcpu id='9' enabled='no' hotpluggable='yes'/>
    <vcpu id='10' enabled='no' hotpluggable='yes'/>
    <vcpu id='11' enabled='no' hotpluggable='yes'/>
    <vcpu id='12' enabled='no' hotpluggable='yes'/>
    <vcpu id='13' enabled='no' hotpluggable='yes'/>
    <vcpu id='14' enabled='no' hotpluggable='yes'/>
    <vcpu id='15' enabled='no' hotpluggable='yes'/>
  </vcpus>
  <cputune>
    <vcpupin vcpu='0' cpuset='1'/>
    <vcpupin vcpu='1' cpuset='2'/>
    <vcpupin vcpu='2' cpuset='3'/>
    <vcpupin vcpu='3' cpuset='4'/>
    <emulatorpin cpuset='5-6'/>
  </cputune>
  <cpu mode='custom' match='exact' check='full'>
  </cpu>

```

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 10 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go:2035]: great idea! 
[DONE](https://github.com/kubevirt/kubevirt/compare/c919c2a4cbb6b8026f1ab23c8e43a2d9acfa0043..fbdee3924ee7e34c214389e3105e791e1b1adf08)

**Review by enp0s3 — APPROVED** [review]

**Comment by vladikr** [comment]
/approve
Thanks.

**Comment by orelmisan** [comment]
> /cherry-pick v1.1.0

The automated cherry-pick is expected to fail because #10757 was not merged to `release-1.1`.

Please consider adding the commits from this PR to #10757 in order for the whole fix to be backported as a single PR.
